### DISCUSSION TITLE
Submit a form outside `onSubmit`

### DISCUSSION THREAD

**Discussion by defrex** [mantinedev/mantine#2784] [Feature requests]
I have a form which required a submit button to be outside the `form` component. Using a ref to call `formRef.current.submit()` skips the `onSubmit` handler in React. `formRef.current.requestSubmit()` exists, but is very new and doesn't work in some browsers (Mobile Safari < 16 is my problem).

It'd be great if I could submit the form using only the `form` object returned from `useForm`. A simple `form.submit()` would be great.

**Comment by defrex** [comment]
Hm, I found a solution using `formRef.current.dispatchEvent(new Event('submit', { bubbles: true, cancelable: true }));` for browsers that don't support `requestSubmit`.

```ts
  const handleSubmitClick = useCallback(() => {
    if (formRef.current === null) {
      return;
    }
    if ('requestSubmit' in formRef.current) {
      formRef.current.requestSubmit();
    } else {
      // Stop TS from treating the feature check as a type guard
      (formRef.current as HTMLFormElement).dispatchEvent(
        new Event('submit', { bubbles: true, cancelable: true }),
      );
    }
  }, []);
```

A manual submit would still probably be handy, but is less necessary than I first thought.

**Comment by rtivital** [comment]
You do not need to submit form in this case: 
- ensure that all values are valid `const valid = form.validate()`
- run your action `valid && yourAction(form.values)`

**Reply by defrex** [reply]
Hah, fair enough! I somehow assumed that the submit handler was doing something more than that, but I can't now think what it would be.

**Reply by Rigel772** [reply]
Is it possible to `transformValues` similar to `form.validate()`?
Thanks for great library!

**Reply by rtivital** [reply]
You can call `form.onSubit` without form event in latest version: `form.onSubmit(yourAction)()`

**Reply by Rigel772** [reply]
Fantastic! Thank you.

**Comment by gius** [comment]
@rtivital what if the [onSubmit](https://github.com/mantinedev/mantine/blob/master/src/mantine-form/src/types.ts#L48) function had a generic type for the event source instead of hardcoded `HTMLFormElement`? We could then use the function for other events as well: `<Button onClick={form.onSubmit(...)}`

I mean:
```
export type OnSubmit<Values, TransformValues extends _TransformValues<Values>, TSourceElement extends HTMLElement> = (
  handleSubmit: (
    values: ReturnType<TransformValues>,
    event: React.SyntheticEvent<TSourceElement>
  ) => void,
  handleValidationFailure?: (
    errors: FormErrors,
    values: Values,
    event: React.SyntheticEvent<TSourceElement>
  ) => void
) => (event?: React.SyntheticEvent<TSourceElement>) => void;
```

...or maybe the whole event type should be a generic param.

**Reply by rtivital** [reply]
The single purpose of event is to prevent form default behavior, there is no need to prevent button default behavior. Just call the function, it will have the same results.

**Reply by gius** [reply]
OK, I understand. 

...and one more idea - could Mantine support async `handleSubmit`? 

I.e.,

```
  handleSubmit: (...) => Promise<void> | void,
```

**Reply by rtivital** [reply]
No, it is not planned, you need to handle async operations on your side
